SCClassLibrary/Common/Core/Function.sc

```diff
@@ -159,6 +159,18 @@ Function : AbstractFunction {
 		^thisThread;
 	}
 
+	whenTrueWithin { |timeout = 3, thenFunc, elseFunc, dt = (1/16)|
+		var remaining = timeout;
+		^Routine {
+			while {
+				remaining = remaining - dt;
+				this.value.not and: { remaining >= 0 }
+			} {
+				dt.wait;
```

Comment:
Two points

First, Condition variables _exist_ so that a thread can be woken when a condition is true. There's no need to reinvent the wheel.

Second, the way this is timed means that the advertised duration is almost never correct:

```supercollider
(
var main_wait = 2;

{ false }.whenTrueWithin(main_wait, nil, { "whenTrueWithin".postln });

15.do { |i|
	r {
		var time = main_wait + (i / 100);
		time.wait;
		"done (%)\n".postf(time);
	}.play
}
)
```

When `main_wait` is 2, I observe a drift of about .04 seconds relative to the routines using a single `wait`. When `main_wait` is 5, the drift is about .12 seconds. This is a minor point to demonstrate how easy it is to make mistakes in multithreaded/time-dependent code; the first point is much more important.

In SC, our threading/async utilities are already fragmented, and it's hard to pin down standard practices. I believe we have now three or four idioms for doing the exact thing this method tries to do. I think the proposal Julian and I had for a wait with timeout in the ML thread was a much better approach; it seems like you're completely ignoring that conversation with this PR, and that doesn't do anyone good.

Reply:
@brianlheim - Sorry I missed the ML thread - which one is that exactly? 
I was aware of the "testing async code" thread on sc.dev, but that does not have a full hangWithTimeout proposal, and #3300 was closed in Nov. 
- is there an open hangWithTimeout proposal that I am not finding on github?

Reply:
Yes, that's the thread, and i was referring to the discussion about waiting-with-timeout on a Condition

Reply:
OK, I read the thread, and did not see a canonical example how to use Condition to wait with timeout - is there one? Or better, a final implementation of hangWithTimeout or similar?
I doubt that I understand Condition well enough to write that myself.
Happy to use it when someone qualified has written it.

> I believe we have now three or four idioms for doing the exact thing this method tries to do. 

Which ones are you thinking of? I know of a few places where such mechanisms are handcoded within different methods, with variations. That is exactly why I thought it makes sense to replace all these with a single method that does this well. 

Ideally, I like to have a single recommended Condition solution for syncing back to a thread, and a single recommended polling solution for simple branching by truth or timeout (like the PR title says), with no need to pass the condition object around - with polling, you can really test whether ANY given test (or set of tests) is met within a given time or not.

PS: The timing drift happens because the fork plays on the AppClock (so gui code can be used in the result functions) - with SystemClock, drift is within 0.01. Can add a clock argument if desired.

Reply:
If you feel you are unqualified to work on Condition, I would question why you feel submitting this PR is a good idea. This is the same problem domain. Perhaps it would be better to focus on other fixes that you are more confident in. It's good that you are identifying a place where our code could be made clearer, but I strongly feel this is not a viable solution. I also do not feel it is incumbent on me (or any other maintainer) to explain from first principles why that is the case, any more than I already have. I just don't have time to do that in every situation, and there are issues that need greater attention. Rather, it is your responsibility to have knowledge of the problem domain before you begin programming in it.

I don't mean this as gatekeeping; there's a reason I don't submit fixes to the server or Patterns library -- I don't understand those components well and I know someone else can fix it better and faster, just like I can fix issues on other components more efficiently. I'm simply asking that we recognize some maintainers are more qualified that others to push fixes in specific areas, and that the project is healthier when we operate under that knowledge. I think continuing to work on the other PRs you have open, while voicing your desire for a solution that looks something like this, is a good compromise.

Reply:
As far as I remember, Alberto has been working on (roughly) this problem domain for the last 15 years. At least then we wrote a paper together on the topic of waiting. Three well versed supercollider developers have discussed this method via email for two weeks before submitting a pull request, so it is already in a good state before it is suggested and hopefully received with a joyful welcome. 

@brianlheim could it be that you haven't quite understood why polling is used here?

When you pass an *arbitrary* function to a method, and you want to know when it becomes true (without any specific arguments passed), what are the alternatives to polling? Remember, you don't have any connection to the process that makes it come true.

Of course, you could argue that this kind of function should not exist, or if it exists it should never be passed to a method called `whenTrueWithin` -- that would be a matter of how you think processes should interact. But the intent of the method is precisely to simplify a coupling of uncoupled processes. 

Because when they are already coupled (say, because the function internally uses a `forkIfNeeded` and a condition), then you don't need such a method in the first place, because you can just wait for it to finish in a surrounding thread.